ethers-providers/src/pending_escalator.rs

```diff
@@ -0,0 +1,198 @@
+        *$this.state =
+            PendingStates::Sleeping(Box::pin(tokio::time::sleep(*$this.polling_interval)));
+        $cx.waker().wake_by_ref();
+        return Poll::Pending;
+    };
+}
+
+macro_rules! completed {
+    ($this:ident, $output:expr) => {
+        *$this.state = Completed;
+        return Poll::Ready($output);
+    };
+}
+
+macro_rules! broadcast_checks {
+    ($cx:ident, $this:ident, $fut:ident) => {
+        match $fut.as_mut().poll($cx) {
+            Poll::Ready(Ok(pending)) => {
+                $this.sent.push(*pending);
+                check_all_receipts!($cx, $this);
+            }
+            Poll::Ready(Err(e)) => {
+                completed!($this, Err(e));
+            }
+            Poll::Pending => return Poll::Pending,
+        }
+    };
+}
+
+impl<'a, P> Future for EscalatingPending<'a, P>
+where
+    P: JsonRpcClient,
+{
+    type Output = Result<TransactionReceipt, ProviderError>;
+
+    fn poll(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        use PendingStates::*;
+
+        let this = self.project();
+
+        match this.state {
+            Initial(fut) => {
+                broadcast_checks!(cx, this, fut);
+            }
+            Sleeping(fut) => {
+                if fut.as_mut().poll(cx).is_ready() {
+                    // if timer has elapsed (or this is the first tx)
+                    if this.last.is_none()
+                        || this.last.clone().unwrap().elapsed() > *this.broadcast_interval
+                    {
+                        // then if we have a TX to broadcast, start
+                        // broadcasting it
+                        if let Some(next_to_broadcast) = this.txns.pop() {
+                            let fut = this.provider.send_raw_transaction(next_to_broadcast);
+                            *this.state = BroadcastingNew(fut);
+                            cx.waker().wake_by_ref();
+                            return Poll::Pending;
+                        }
+                    }
+
+                    check_all_receipts!(cx, this);
+                }
+
+                return Poll::Pending;
+            }
+            BroadcastingNew(fut) => {
+                broadcast_checks!(cx, this, fut);
+            }
+            CheckingReceipts(futs) => {
+                // if drained, sleep
+                if futs.is_empty() {
+                    sleep!(cx, this);
+                }
+
+                // otherwise drain one and check if we have a receipt
+                let mut pollee = futs.pop().expect("checked");
+                match pollee.as_mut().poll(cx) {
+                    //
+                    Poll::Ready(Ok(Some(receipt))) => {
+                        completed!(this, Ok(receipt));
+                    }
+                    // rewake until drained
+                    Poll::Ready(Ok(None)) => cx.waker().wake_by_ref(),
```

Comment:
Is this `cx.waker().wake_by_ref()` necessary? Won't executor automatically put the EscalatingPending future back in queue? Other paths here that return Poll::Pending don't call wake

Reply:
the goal is to explicitly notify the executor that this is ready to be polled again _immediately_. i.e. override the executor's scheduling and jump to the front of the queue as much as we have the ability to, as we know that there's no async operation happening to wait on

Reply:
this section looks like it could be wrapped in a loop over all futures? So that `Poll::Ready(Ok(None)) =>` merely results in a continue?
Is the order in which they should be polled here important, if not then `Vec::swap_remove` could be useful.

Reply:
I mean it kinda is right now, it's just ensuring that it allows the executor to schedule each step of the loop and interleave other work

Reply:
Looping over all futures is an anti-pattern I think, the state machine _may_ be slightly easier to write but as @prestwich says it's better to do it on a per-fut basis

Reply:
Isn't the `CheckingReceipts` state essentially a `FuturesOrdered` but we only poll one future at a time?